src/autoscaling/snapshotter.js

```diff
@@ -0,0 +1,267 @@
+ *
+ * Memory becomes overloaded if its current use exceeds the `minFreeMemoryRatio` option.
+ * It's computed using the total memory available to the container when running on
+ * the Apify platform and a quarter of total system memory when running locally.
+ * Use of total memory may be overridden by using the `maxBlockedRatio` option.
+ *
+ * Event Loop becomes overloaded if it slows down by more than the `maxBlockedMillis` option.
+ *
+ * CPU tracking is available only on the Apify platform and the CPU overloaded event is read
+ * directly off the container and is not configurable.
+ *
+ * @param {Object} options
+ * @param {Number} [options.eventLoopSnapshotIntervalSecs=0.5]
+ *   Defines the interval of measuring the event loop response time.
+ * @param {Number} [options.maxBlockedMillis=50]
+ *   Maximum allowed delay of the event loop in milliseconds.
+ *   Exceeding this limit overloads the event loop.
+ * @param {Number} [options.memorySnapshotIntervalSecs=1]
+ *   Defines the interval of measuring memory consumption.
+ *   The measurement itself is resource intensive (25 - 50ms async),
+ *   therefore, setting this interval below 1 second is not recommended.
+ * @param {Number} [options.maxUsedMemoryRatio=0.7]
+ *   Defines the maximum ratio of memory that can be used.
+ *   Exceeding this limit overloads the memory.
+ * @param {Number} [options.snapshotHistorySecs=60]
+ *   Sets the interval in seconds for which a history of resource snapshots
+ *   will be kept. Increasing this to very high numbers will affect performance.
+ * @ignore
+ */
+export default class Snapshotter {
+    constructor(options = {}) {
+        const {
+            eventLoopSnapshotIntervalSecs,
+            memorySnapshotIntervalSecs,
+            snapshotHistorySecs,
+            maxBlockedMillis,
+            maxUsedMemoryRatio,
+            maxMemoryMbytes,
+        } = _.defaults(options, DEFAULT_OPTIONS);
+
+        checkParamOrThrow(eventLoopSnapshotIntervalSecs, 'options.eventLoopSnapshotIntervalSecs', 'Number');
+        checkParamOrThrow(memorySnapshotIntervalSecs, 'options.memorySnapshotIntervalSecs', 'Number');
+        checkParamOrThrow(snapshotHistorySecs, 'options.snapshotHistorySecs', 'Number');
+        checkParamOrThrow(maxBlockedMillis, 'options.maxBlockedMillis', 'Number');
+        checkParamOrThrow(maxUsedMemoryRatio, 'options.maxUsedMemoryRatio', 'Number');
+        checkParamOrThrow(maxMemoryMbytes, 'options.maxMemoryMbytes', 'Maybe Number');
+
+        this.eventLoopSnapshotIntervalMillis = eventLoopSnapshotIntervalSecs * 1000;
+        this.memorySnapshotIntervalMillis = memorySnapshotIntervalSecs * 1000;
+        this.snapshotHistoryMillis = snapshotHistorySecs * 1000;
+        this.maxBlockedMillis = maxBlockedMillis;
+        this.maxUsedMemoryRatio = maxUsedMemoryRatio;
+        if (maxMemoryMbytes) this.maxMemoryBytes = maxMemoryMbytes * 1024 * 1024;
```

Comment:
It's a good practice to always defined the exactly same object props, even if they are null. It's also important for better V8 optimizations. So I'd change this to:
```
this.maxMemoryBytes = maxMemoryMbytes ? maxMemoryMbytes * 1024 * 1024 : null;
```